### ISSUE TITLE
Add funsor.numpy backend

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/funsor#1]
Currently `funsor.torch` provides a PyTorch backend via `Tensor`, `@function` and a few other helpers. It should be straight-forward to provide similar adaptors for numy and jax., say as a new module `funsor.numpy` and class `Array`.

## Outline

- [x] #74 support numpy in `pyro.ops`. e.g. we will need to call `np.log(x)` rather than `x.log()` based on type.
- [x] implement an `align_arrays` function similar to `align_tensors`
- [x] implement a `funsor.numpy.Array` class
- [x] register `np.array` with `to_funsor`
- [x] fork `test_torch.py` to a new `test_numpy.py`
- [x] Add compatibility with tail-call optimized interpreter from #145 and remove conditional xfail added to `test_numpy.py` added in #155 
- [x] implement `Array.eager_reduce()`
- [x] optionally add a `Function` wrapper and `@funsor.numpy.function` decorator
- [x] refactor (`Tensor`,`Array`) to share a superclass, e.g. `Tensor` -> {`TorchTensor`, `NumpyTensor`)
- [x] refactor `Gaussian` to `Tensor` rather than torch.Tensor internally.
    This requires wrapping many linear algebra operations as functions, e.g. `cholesky_solve`.
- [x] Make unary ops work with Array input. (#303)
- [ ] Enable `test_tensor.py` with FUNSOR_USE_TCO=1
- [x] Support numpy docstring in `funsor.util.getargspec`
- [x] Add numpy backend tests for `test_gaussian`.
- [x] Add numpy backend for `einsum`.
- [ ] Consider disabling some of the overhead caused by `normalize`.
- [x] Make `funsor.distributions` backend-agnostic
- [x] Add support for `jax.numpy`.
- [ ] update `funsor.minipyro` to support numpy backend
- [x] add inference examples using numpy

**Comment by eb8680** [comment]
> Enable test_tensor.py with FUNSOR_USE_TCO=1

@fehiepsi this should currently work with PyTorch but not NumPy.  Unless there was a regression in #303 and the PyTorch version is failing too, there's no need to prioritize this.

**Comment by eb8680** [comment]
I'm going to close this as resolved. We can track progress on `normalize` in #411.
